### ISSUE TITLE
The app is not remembering me

### ISSUE THREAD

**Issue by Foxandxss** [heartsentwined/ember-auth-rails-demo#3]
Hi, I tried both the heroku app and I cloned the repo to try myself but I log in (with remember_me checked), I close the browser, open it again and I'm not logged in anymore.

Thanks :)

**Comment by heartsentwined** [comment]
Remember Me will only be triggered (as is the default) when you actually visits a protected route. In this case:
- `users.*` are protected (auto sign in here)
- `posts.*` are not protected (no auto sign in)
- `index` / "home page" is not protected (no auto sign in)

Try entering http://ember-auth-rails-demo.herokuapp.com/users directly after you have signed in (with remember me), and you should see the session remembered.

**Comment by heartsentwined** [comment]
heartsentwined/ember-auth#36 has a full discussion on this default, and on how to trigger the (optional) behavior of auto-recall session for public routes (in addition to auth-only routes).

**Comment by Foxandxss** [comment]
Oh, that works, I just find weird that defaults (Normally you're auto-signed even if you don't move to a protected place). Anyways great work :) Discovered a lot of things in the tutorial about rails (Never used that token_authenticatable)

**Comment by heartsentwined** [comment]
cf. the issue I mentioned. This "default" depends on whether you are using a site heavy on protected content (a la github, facebook), or heavy on public content (banking sites).

**Comment by AshesOfOwls** [comment]
I wanted to add to this as I am still having a related issue.

Currently if I have logged in and have remember me checked and I visit this page: http://ember-auth-rails-demo.herokuapp.com/users

None of the users in the users list load in right away. Instead I have to go to a separate page and go back to /users before it authenticates correctly. It seems to be an issue with ember not authenticating before it requests /api/users.json and therefore rails still sees the user as not logged in.

**Comment by Foxandxss** [comment]
That's true, the user list doesn't load if you go to that url directly and you're already logged in.

**Comment by happyFish** [comment]
@Foxandxss , is there a way to trigger the route to get the model once the user has been auto signed in?

**Comment by Foxandxss** [comment]
@happyFish that question belongs to @heartsentwined :P

**Comment by heartsentwined** [comment]
**TL;DR**: the current implementation is flawed; it just "won't work", not even with the `signedIn` conditional check. No "full" solution yet.

A rather thorny in ember development for now. I have been tracing down the issue and a not-too-verbose solution for weeks, here's what I've got so far:
- We can't just return nothing on `model` calls. The standard ember-data `find()` calls returns a skeleton / dummy _immediately_, before your API returns anything, so that the view layer (and other ember components) will have something to work with, and, more importantly, so that they can just call the model methods as if the model is already there. It is, in essence, an empty record with absolutely no data in it.
- ember-data keeps track of this dummy model, and it injects real data in it when the API response comes back. ember's bindings, in turn, auto-updates the view layer (and computed properties, etc) with this data.
- However, some hooks are called only once, e.g. `setupController()`. It won't re-trigger when "real" data comes back; it will only trigger on the first "dummy model" initialization.

Where does this leave us?
- The best solution is to make an, e.g. `App.Secret`, proxy, register it with `ember-data`, keep track of it, issue our own `find()` on model when appropriate (i.e. signed in), and then populate the proxy. Take care also to prevent ember / ember-data itself from registering two instances of the model (once from our proxy, a second time from the `find()`), and take care to reify the `find()` for our particular model ID for ember-data, and make it point to our proxy (now real model). - That is, if someone can dig deep into ember-data and implement this chain for us, we will all be grateful. :smile:
- For now, merely observing the model and re-calling `model()` won't work, because after sign in, the newly `find()` model won't have any link with the view layer (and other components) whatsoever.
- The conditional won't work, again because the view layer is filled with _nothing_ to begin with.
- Reliance on standard hooks will need to be re-worked somehow, because you might expect the hooks to get triggered after a user signs in? (They won't.)
- Same for custom logic. You could make the observer call all those hooks after sign in, but that's verbose, prone to error (forgetting one hook for one particular route), and in any case calling some elementary hooks (e.g. `create()` / `init()`) _might_ entail some ember-default actions that we don't want re-triggering?

**Status**? - Still searching for a solution. :disappointed:

**Comment by heartsentwined** [comment]
Reopening until resolved.

**Comment by Globegitter** [comment]
Is there any update? Do you think there will be a solution to that problem soon. I actually sent you an email a few hours ago, because I started looking into ember-auth and this is be a critical future, which could change some of the decisions we make. So any sort of further update or news would be brilliant.

**Comment by heartsentwined** [comment]
Not much `ember-auth` issue, but `ember-data` is not behaving. I already have to send two PRs (emberjs/data#994, emberjs/data#999), also refer to emberjs/data#724. Status? I'm still ironing out the edges (specifically on `embedded` behavior). If you like, I can upload my WIP in a repo, but I'm not sure it would create more questions / wrong guidances or solutions for you.

**Comment by Globegitter** [comment]
So is this issue dependent on ember-data including all these 3 PRs? Or is that something you can 'fix' in ember-auth? I suppose this also partially answers my next question: Is your WIP just an update to ember-auth or are you also using a forked/modified version of ember-data?
Just asking so I can get a clearer picture of what is going on and think about next steps. The help is really appreciated.

**Comment by heartsentwined** [comment]
Not sure at the moment.

I'm using a forked version of `ember-data` at [heartsentwined/ember-data](https://github.com/heartsentwined/ember-data): emberjs/data#994 can be implemented via a patch in `ember-auth`, but emberjs/data#999 seems to be only implementable via a direct patch of `ember-data`. In any case, I believe that both are useful patches to `ember-data`, so I'm submitting them as PRs for the community anyway.

Then there is the hack to get over authenticated models - ref emberjs/ember.js#2041. That might make it back to `ember-auth` as official "patched" router / routes API, but then there are still proposals to revamp the router / routes at ember side. The latest seems to be [this facelift](https://gist.github.com/machty/5723945), which addresses our problems here directly. If they implement this, then obviously we wouldn't want a "transitional" patched API here in `ember-auth`.

emberjs/data#724 was used in an earlier iteration of my attempt, but then it fails in atomicity guarantee - in addition to still having bugs -, so it is dropped now; but I'm putting it here for future reference.

I'll try to doc my WIP in a repo, so that more eyes can see what's going on, and hopefully more brains can get involved to solve the issues.

**Comment by Globegitter** [comment]
@heartsentwined the new rc6 has the facelift you mentioned included right? (The changelog states: \* Router facelift – Async transitions) So we are a step closer to fixing this if I understand it right. That seems like some good news :)

**Comment by Foxandxss** [comment]
Yep, that facelift is there in ember rc6. So I guess that ember-auth codebase can be really improved and address some nasty bugs. @heartsentwined will do his magic to gives us the best auth solution out there :)

**Comment by heartsentwined** [comment]
@Foxandxss that's very flattering :smile: But seriously, as you have observed, the new router is heavily relevant to authenticated-backed apps. I'm working with it to update ember-auth with the new APIs. For now, this issue is put on hold - I'll reproduce the scenario after I finish with the new router, and see if the issue persists.

Ref:
- [gist](https://gist.github.com/machty/5723945)
- [presentation](https://machty.s3.amazonaws.com/ember-facelift-presentation/index.html)

**Comment by simpleshadow** [comment]
@heartsentwined No pressure, but any rough idea how long until you have ember-auth updated to rc 6? Any help needed? Thanks for all your work so far. I've recently started adding ember-auth to our project but it seems futile to continue if you're working to integrate rc 6 as we too are upgrading.

**Comment by Foxandxss** [comment]
Yeah, with this kind of route revamp, a big work is needed but it will result in a better library :)

**Comment by heartsentwined** [comment]
@simpleshadow you know, one can _never_ predict how long coding would take

**Comment by heartsentwined** [comment]
`7.x` branch now released, with rc6 (new router!) support. [changelog](https://github.com/heartsentwined/ember-auth/blob/master/CHANGELOG.md), [upgrade guide](https://github.com/heartsentwined/ember-auth/blob/master/CHANGELOG.md#upgrade-guide), [docs](http://ember-auth.herokuapp.com/docs) and [demo](https://github.com/heartsentwined/ember-auth-rails-demo/) all properly updated. Main notes:
- `ember-auth` now _requires_ ember `rc6` or above
- the new router now _encourages_ you to use `authRedirectable` as "the ember way" of solving auth-only routes (with auth-only models)
- thus you might want to pair it with `actionRedirectable` too for user experience
- despite this being a major version update, it would likely not break most userland code; the BC breaks are not in main, documented API

---

Closing since the [reopened issue](https://github.com/heartsentwined/ember-auth-rails-demo/issues/3#issuecomment-16876407) described by @McPants has gone away in the upgraded demo.
